src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionContext.cs

```diff
@@ -124,12 +157,61 @@ public override ValueTask<ConnectionContext> ConnectAsync(IFeatureCollection? fe
                 quicStream = _connection.OpenBidirectionalStream();
             }
 
-            var context = new QuicStreamContext(quicStream, this, _context);
+            // Only a handful of control streams are created by the server and they last for the
+            // lifetime of the connection. No value in pooling them.
+            QuicStreamContext? context = new QuicStreamContext(this, _context);
+            context.Initialize(quicStream);
             context.Start();
 
             _log.ConnectedStream(context);
 
             return new ValueTask<ConnectionContext>(context);
         }
+
+        internal bool TryReturnStream(QuicStreamContext stream)
+        {
+            lock (_poolLock)
+            {
+                if (!_streamPoolHeartbeatInitialized)
+                {
+                    // Heartbeat feature is added to connection features by Kestrel.
+                    // No event is on the context is raised between feature being added and serving
+                    // connections so initialize heartbeat the first time a stream is added to
+                    // the connection's stream pool.
+                    var heartbeatFeature = Features.Get<IConnectionHeartbeatFeature>();
+                    if (heartbeatFeature != null)
```

Comment:
In [HttpConnection](https://github.com/dotnet/aspnetcore/blob/12c57a5e0ae5bc883e3ff46af90887b2965d5563/src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs#L93-L101) we assert the feature exists. I think we should be even more aggressive here and blow up. It's too bad we don't have a better place to initialize this, but I see why it can't be done much earlier.